### PR TITLE
Support Ethereum pubkeyhash

### PR DESCRIPTION
**Pull request by axic**
Based on the discussion: https://github.com/trezor/trezor-common/pull/11

Comments welcome

### CODE REVIEW

**Comment by prusnak** [comment]
Please implement `uncompress_pubkey` function by using `ecdsa_read_pubkey` and then serializing `curve_point` structure back to `(uint8_t *)` using bn_write_be. It will be cleaner, shorter and that way it will also use `ecdsa_validate_pubkey`.

**Comment by axic** [comment]
A quick note: there's an `uncompress_coords` method which should be made private or `ecdsa_` prefixed too.

**Comment by axic** [comment]
Changed (in any case don't merge, want to squash some commits later).

One downside I see is the speed penalty when an uncompressed key is passed, that however is not going to happen in the firmware as everything is compressed in BIP32 and the firmware won't uncompress it. Correct?

**Comment by prusnak** [comment]
Where exactly is this speed penalty when dealing wit uncompressed keys?
 
Right, bip32 always works with compressed keys.
 
--
Best Regards / S pozdravom,
 
Pavol Rusnak <stick@gk2.sk>
 
 
On Tue, May 17, 2016, at 08:44 PM, Alex Beregszaszi wrote:
> Changed (in any case don't merge, want to squash some commits later).
> One downside I see is the speed penalty when an uncompressed key is
> passed, that however is not going to happen in the firmware as
> everything is compressed in BIP32 and the firmware won't uncompress
> it. Correct?
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub[1]
>
 

Links:

  1. https://github.com/trezor/trezor-crypto/pull/59#issuecomment-219814658

**Comment by axic** [comment]
> Where exactly is this speed penalty when dealing wit uncompressed keys?

The previous code was copying the input if it was uncompressed. The new one does (de)serialisation and verification.  I like verification better though.

> Right, bip32 always works with compressed keys.

BIP32 does. So does Trezor in the entire codebase?

**Comment by prusnak** [comment]
> The previous code was copying the input if it was uncompressed. The new one does (de)serialisation and verification. I like verification better though.

Ah, now I understand.

> BIP32 does. So does Trezor in the entire codebase?

Yes.

File: ecdsa.c [hunk 1]
```diff
@@ -829,6 +849,30 @@ void ecdsa_get_pubkeyhash(const uint8_t *pub_key, uint8_t *pubkeyhash)
 	MEMSET_BZERO(h, sizeof(h));
 }
 
+#if USE_ETHEREUM
+int ecdsa_get_ethereum_pubkeyhash(const uint8_t *pub_key, uint8_t *pubkeyhash)
+{
+	uint8_t h[65];
+	SHA3_CTX ctx;
+
+	if (!ecdsa_uncompress_pubkey(&secp256k1, pub_key, h)) {
```
Comment (Arachnid) [inline ecdsa.c]: If this returns 0, `h` will be unmodified so the memset below is redundant, but `pub` in `ecdsa_uncompress_pubkey` may contain unzeroed data.

Comment (axic) [inline ecdsa.c]: True about `h`.

Actually if you look up the chain, I believe it is more the duty of `ecdsa_read_pubkey` to clean `pub` if `ecdsa_validate_pubkey` fails. The same issue is present in `ecdsa_verify_digest`. @prusnak, what do you think?

File: ecdsa.c [hunk 2]
```diff
@@ -36,6 +36,9 @@
 #include "base58.h"
 #include "macros.h"
 #include "secp256k1.h"
+#ifdef USE_ETHEREUM
```
Comment (jhoenicke) [inline ecdsa.c]: This should be #if USE_ETHEREUM.

Or change options.h to not define USE_ETHEREUM/USE_KECCAK to 0 and change all #if to #ifdef.

Comment (axic) [inline ecdsa.c]: This should be #if to be in line with the rest of the code.

File: ecdsa.c (hunk 1 above)
Comment (axic) [inline ecdsa.c]: @prusnak @jhoenicke can you please review the previous comment? I will be able to push an update in the late afternoon once agreed on this.

Comment (prusnak) [inline ecdsa.c]: `ecdsa_uncompress_pubkey`, `ecdsa_read_pubkey`, `ecdsa_verify_digest`, `ecdsa_validate_pubkey` etc. all deal with public data. There is no need to clear stack variables. I suggest just removing MEMSET_BZERO calls from line 859 and 870 below.

**Comment by jhoenicke** [comment]
Maybe you can add some small test case for this, e.g. to test.c.  Just feed it a few compressed and uncompressed pubkeys (both 02 and 03 and even/odd 04) and check against a known output.

My only concern is that `ecdsa_uncompress_pubkey` basically just wraps `ecdsa_read_pubkey` giving the point back in serialized 04... form and then the only caller `ecdsa_get_ethereum_pubkeyhash` even ignores the 04.  Maybe it is simpler to inline the function.  But that is my personal taste.

There is also `ecdsa_get_public_key65`, but this needs a private key to work.  This function is only used by the new u2f code.

File: ecdsa.c (hunk 1 above)
Comment (axic) [inline ecdsa.c]: Understood.  You are correct the only reason I kept it for public data because it was present in the majority of the code.

**Comment by axic** [comment]
> My only concern is that ecdsa_uncompress_pubkey basically just wraps ecdsa_read_pubkey giving the point back in serialized 04... form and then the only caller ecdsa_get_ethereum_pubkeyhash even ignores the 04. Maybe it is simpler to inline the function.

That might be an option. Ethereum only uses the x/y points and not the SEC1 serialised form.

> There is also ecdsa_get_public_key65, but this needs a private key to work.

By looking at mcu/fsm.c, the HDNode should have the private key in every single case available.  We could use a method based on `ecdsa_get_public_key65`. What would you name such a method?

**Comment by jhoenicke** [comment]
Not sure about naming, but something like `get_ethereum_address` or the current `...pubkeyhash` is okay.  In the long run I would like to have all such the function in bip32, taking only a `hdnode` (which probably then uses functions in `ecdsa.c`).  Then one can easily change the implementation to use the private key, without changing the interface.

At least for multisig signing (I'm not sure if that exists in Ethereum), we support `hdnode`s without a private key.

File: ecdsa.c (hunk 1 above)
Comment (axic) [inline ecdsa.c]: This was changed - the inline comment refers to a different line, hence this thread isn't collapsed by Github.

**Comment by axic** [comment]
@jhoenicke @prusnak added tests for both new methods

**Comment by prusnak** [comment]
Thanks!

**Pull request merged**